### DISCUSSION TITLE
bug: no type safety when using nested routers?

### DISCUSSION THREAD

**Discussion by jclaessens97** [trpc/trpc#2933] [Q&A] [answered]
### Provide environment information

```
❯ npx envinfo --system --binaries --browsers --npmPackages "typescript,next,react,@trpc/server,@trpc/client,@trpc/react,@trpc/client,@trpc/next,@tanstack/react-query"
Need to install the following packages:
  envinfo
Ok to proceed? (y) y

  System:
    OS: macOS 12.4
    CPU: (10) arm64 Apple M1 Max
    Memory: 16.16 GB / 64.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.14.0 - ~/.asdf/installs/nodejs/16.14.0/bin/node
    npm: 8.3.1 - ~/.asdf/plugins/nodejs/shims/npm
  Browsers:
    Chrome: 105.0.5195.125
    Safari: 15.5
```

### Describe the bug

I followed a simple example on how to configure a multi router setup (using the child routers). However when I'm then trying to use it in my client, I only have limited type support. It only recognizes the first "layer" of the router (being the child router). On the childrouters itself I don't seem to have any typings.

Might be that I'm doing something wrong, but I followed the example very closely.

~Another thing is that I'm working with a separate server, but I have to import the AppRouter in my web project directly from the API project. Is there another, cleaner way, of doing this?~
EDIT: I just read the docs about only importing the type 

### Link to reproduction

https://github.com/jclaessens97/trcp-nested-route-repro

### To reproduce

Implement nested router, use it in client, see that type safety stops at the first layer.

### Additional information

versions:
```
    "@trpc/client": "10.0.0-proxy-beta.15",
    "@trpc/server": "10.0.0-proxy-beta.15",
```

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by KATT** [comment, marked as answer]
On my phone but this is not a bug in trpc. Might be a bad alias here? 
![Screenshot_2022-10-07-14-11-15-26_40deb401b9ffe8e1df2f1cc5ba480b12](https://user-images.githubusercontent.com/459267/194561929-e5cc86fa-ad12-4047-a31c-ebdd5ec47612.jpg)

**Reply by Alexnortung** [reply]
I'm new to trpc, and have a similar issue, what was the solution here?

**Comment by jclaessens97** [comment]
Good catch! Sorry about that 😅

**Comment by talksik** [comment]
Same issue, I believe that this is related to `https://github.com/microsoft/TypeScript/issues/23918`

Importing a deeply nested type results in typescript just inferring that certain fields are of type `any`.

**Reply by talksik** [reply]
This happens when I have a nested router and the file that defines the composite router is importing other routers and merging them. Having them in the same file resolves it and type script correctly infers the nested router types properly.

For context, I have a monorepo setup with server and frontend react app as separate packages.

**Reply by talksik** [reply]
However, it's annoying because the return type of nested routers handlers is inferred by typescript as `any`

To workaround this, I put explicit return types for each handler like this so that the frontend can more easily infer the return type.

```
getMessages: procedure.input(...).query() : Promise<string> {
return 'test'
}
```

**Reply by kyle-belle** [reply]
I was kinda having the same issue.

The problem in my case was that my for some reason my tsconfig `compilerOptions.baseUrl`  & `compilerOptions.path` doesn't seem to work correctly (but yet vs code still auto-imports them for me from the baseUrl).

i was importing the functions i was using the procedure like this:
```typescript
import { getMessages } from "src/utils/message";
// ...
getMessages: procedure.input(...).query(() => {
return await getMessages();
})
```

but i actually needed to import it using a relative path like this:
```typescript
import { getMessages } from "../../utils/message";
// ...
getMessages: procedure.input(...).query(() => {
return await getMessages();
})
```

**Reply by puerschel93** [reply]
@kyle-belle that worked for me too, is there an actual working workaround for this without relying on  imports like `../../example` ?

**Reply by kyle-belle** [reply]
You should be able to use path aliases in your tsconfig and and the common folders aliased
